elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,420 @@
+resolution of expressions.
+
+The `expression=encryption` resource will contain three attributes, the first attribute being `resolvers` which will be used to define one or more resolvers to
+handle the encryption and decryption of values.  Each resolver will support the following configuration:
+
+ * `name` - Unique name of the resolver, this can be reference back by the expression.
+ * `credential-store` The name of the credential store to use to load the secret key.
+ * `secret-key` - The alias of the secret key within the credential store.
+
+The format of the expressions resolved using the WildFly Elytron expression resolver will be `${ENC:Resolver:ENCRYPTED_DATA}` where `Resolver` is a reference
+to the specific resolver defined in the `expression=encryption` resource and `ENCRYPTED_DATA` is the data to be decrypted encoded using Base64.
+
+The second attribute on the `expression=encryption` resource will be `default-resolver` which will be used to optionally specify which resolver is the default, the expression can then be simplified to `${ENC:ENCRYPTED_DATA}`.
+
+As expressions are already widely in use with the application server there is a small possibility that users may have already defined expressions with a prefix of `ENC:` expecting it to be resolved as a system property.  The `expression=encryption` resource will also have a third attribute `prefix` attribute which will allow an alternative prefix to `ENC` to be specified.
+
+TODO - I need to double check if the format should be `ENC:` or `ENC::` the former may be easier to accidentially interpret as a system property with a
+default value.
+
+=== Encryption
+
+To simplify the first iteration of this enhancement the only supported transformation will be `AES/CBC/PKCS5Padding`, this has been selected as one of the
+transformations all JVMs are required to support.  Additionally as the values to be encryopted will be of varying lengths padding is required as the values
+may not fit neatly into the block size for AES encryption.
+
+Additional requirements such as the generation of an initialisation vector will be handled by the underlying `Cipher` implementation, different providers
+may choose to handle this differently so we will not second guess their requirements.
+
+==== Command Line Encryption
+
+It was initially considered if a new top level command should be added to the WildFly Elytron Tool to support the encryption of clear text values, however as
+this enhancement is only making use a single transformation we do not have a need for many additional configuration parameters beyond those required to
+initialise and access the underlying credential store.
+
+The `credential-store` command will be updated with an additional action `encrypt`, this in turn will make use of two parameters:
+
+ * `--alias` - The alias of the `SecretKey` stored within the credential store.
```

Comment:
According to the implementation, an alias is taken as the argument of the action `encrypt`. The implementation is consistent.

Reply:
Done